Fix the problem with socket filename size limitation for circus IPC protocol

sphuber opened the pull request:
Fixes #1317 

Currently, we are using the IPC protocol as the default for the daemon to
communicate with the endpoints of the circus controller. The sockets used
to be stored in the AiiDA config folder, however, UNIX has a built in byte
limit of 107 on socket filenames and given that the config folder can be
nested arbitrarily deeply, this limit was easily exceeded. The solution is
to store them in a temporary directory. This solution should always avoid the
limit comfortably but adds slightly more complexity. A temporary folder is
generated by the OS and is only accesible by the user who created it, so
the security is guaranteed here, which is what the IPC protocol was supposed
to solve in the first place over the TCP protocol.

giovannipizzi commented:
Note that on Mac, the temp folder is not /tmp but something longer (even if /tmp exists and could be used, I think):

```
In [2]: tempfile.mkdtemp()
Out[2]: '/var/folders/f4/045f_kfx2wn9w42rjpq68jd00000gn/T/tmpHOofY4'
```
If this is the typical length, this might still be within the limits, unless (if I understand correctly how the filename is generated) the user chooses a profile name that is very long.

Maybe a solution would be, now that there is a UUID identifying the profile, to use the UUID in the filename, that has a fixed length?

Note however that 
`/var/folders/f4/045f_kfx2wn9w42rjpq68jd00000gn/T/tmpHOofY4/circus-bc88d926-435e-47ec-9f43-4367d9df70fa.controller.socket` is already 120-chars long...

I just tested with a script similar to the one I mentioned in the issue, namely:
```python
# -*- coding: utf-8 -*-
import socket
import os, os.path
import time

socket_fname = "/var/folders/f4/045f_kfx2wn9w42rjpq68jd00000gn/T/tmpHOofY4/circus-bc88d926-435e-47ec-9f43-4367d9df70fa.controller.socket"

cutoff_length = 103

socket_fname = socket_fname[:cutoff_length]

print socket_fname

if os.path.exists(socket_fname):
      os.remove(socket_fname)

print "Opening socket..."
server = socket.socket( socket.AF_UNIX, socket.SOCK_DGRAM )
server.bind(socket_fname)

if False:
    print "Listening..."
    while True:
        datagram = server.recv( 1024 )
        if not datagram:
            break
        else:
            print "-" * 20
            print datagram
            if "DONE" == datagram:
                break

print "-" * 20
print "Shutting down..."
server.close()
os.remove(socket_fname)
print "Done"
```
and the limit on Mac OS X is 103 chars, e.g.:
`/var/folders/f4/045f_kfx2wn9w42rjpq68jd00000gn/T/tmpHOofY4/circus-bc88d926-435e-47ec-9f43-4367d9df70fa.`

Maybe one could use a part of the UUID, shorten a bit the name of the filename template, and as a safety measure cutoff the length to 103 chars anyway? 
Another solution would be to put the files in `/tmp` but I don't know if this will always work, and if this is doable with `mkdtemp` (you can specify a prefix, but only within the system temp folder).

sphuber commented:
Ok I did not expect Mac OS being BSD derivative to do something so different. The profile does indeed have a uuid and is 32 bytes (it doesn't use dashes). If we were to take half of the uuid, we would end up with `/var/folders/f4/045f_kfx2wn9w42rjpq68jd00000gn/T/tmpHOofY4/circus-63e78552df76.socket` which is 85 characters. Should that give us enough leeway you think?

greschd commented:
If we're already using a ``tempfile.TemporaryDirectory`` there's not really any need to still make the socket name unique, is there? I think python guarantees that you get a new empty directory anyway. So it could just be named ``circus.sock``.

sphuber commented:
Good point, I have removed the profile dependency of the sockets and reduced their length even further. I have also added a unit test that checks that the returned socket path lengths do not exceed the OS limit which is exposed through `zmq.IPC_PATH_MAX_LEN`

sphuber commented:
Considering point 2) no that is not what is happening. That is why I had to put in the `_SOCKET_DIRECTORY` variable. To know whether a new directory with sockets has to be created or an existing one has to be read, is determined whether the daemon is running. However, I cannot use that check exclusively, because that will only be positive until after the daemon is started and I need to generate the sockets for all three in the same directory. So the first one to call it will create it and set the variable. The call for the two other endpoints will then find the `_SOCKET_DIRECTORY` variable set and use that instead
